drv/sprot-api/src/lib.rs

```diff
@@ -333,232 +422,238 @@ impl From<u8> for MsgType {
     }
 }
 
-#[derive(Copy, Clone, Serialize, Deserialize, SerializedSize)]
-#[repr(C, packed)]
-struct MsgHeader {
-    protocol: Protocol,
-    msgtype: MsgType,
-    len: u16,
+/// A builder/serializer for messages that wraps the transmit buffer
+///
+/// Each public method returns the serialized buffer that can be sent on the
+/// wire.
+pub struct TxMsg {
+    buf: [u8; BUF_SIZE],
 }
 
-#[derive(
-    Copy, Clone, Eq, PartialEq, Serialize, Deserialize, SerializedSize,
-)]
-#[repr(C)]
-pub enum UpdateRspKind {
-    Ok = 1,
-    Value = 2,
-    Error = 3,
-    Unknown = 0xff,
-}
+impl TxMsg {
+    pub fn new() -> TxMsg {
+        TxMsg { buf: [0; BUF_SIZE] }
+    }
 
-impl From<u8> for UpdateRspKind {
-    fn from(kind: u8) -> Self {
-        match kind {
-            kind if kind == (UpdateRspKind::Ok as u8) => UpdateRspKind::Ok,
-            kind if kind == (UpdateRspKind::Value as u8) => {
-                UpdateRspKind::Value
-            }
-            kind if kind == (UpdateRspKind::Error as u8) => {
-                UpdateRspKind::Error
-            }
-            _ => UpdateRspKind::Unknown,
+    pub fn as_slice(&self) -> &[u8] {
+        &self.buf[..]
+    }
+
+    pub fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.buf[..]
+    }
+
+    pub fn payload_mut(&mut self) -> &mut [u8] {
+        &mut self.buf[HEADER_SIZE..BUF_SIZE - CRC_SIZE]
+    }
+
+    /// Serialize an ErrorRsp with a one byte payload, which is a serialized
+    /// `SprotError`
+    pub fn error_rsp(&mut self, err: SprotError) -> VerifiedTxMsg {
+        let payload_size = 1;
+        self.buf[HEADER_SIZE] = err as u8;
+        self.from_existing(MsgType::ErrorRsp, payload_size)
+            .unwrap_lite()
+    }
+
+    /// Serialize a request with no payload
+    pub fn no_payload(&mut self, msgtype: MsgType) -> VerifiedTxMsg {
+        let payload_size = 0;
+        self.write_header(msgtype, payload_size);
+        self.write_crc(payload_size)
+    }
+
+    /// Serialize a request from a MsgType and Lease
+    pub fn from_lease(
+        &mut self,
+        msgtype: MsgType,
+        source: Leased<R, [u8]>,
+    ) -> Result<VerifiedTxMsg, SprotError> {
+        if source.len() > PAYLOAD_SIZE_MAX {
+            return Err(SprotError::Oversize);
         }
+
+        let dest = &mut self.buf[HEADER_SIZE..][..source.len()];
+        source
+            .read_range(0..source.len(), dest)
+            .map_err(|_| SprotError::TaskRestart)?;
+
+        self.write_header(msgtype, source.len());
+        Ok(self.write_crc(source.len()))
     }
-}
 
-impl From<UpdateRspKind> for u8 {
-    fn from(kind: UpdateRspKind) -> Self {
-        match kind {
-            UpdateRspKind::Ok => UpdateRspKind::Ok as u8,
-            UpdateRspKind::Value => UpdateRspKind::Value as u8,
-            UpdateRspKind::Error => UpdateRspKind::Error as u8,
-            _ => UpdateRspKind::Unknown as u8,
+    /// Serialize a request from a buffer with an already written payload
```

Comment:
Clarify that the already-written payload is in **our** buffer, i.e. `self.buf`

Reply:
Fixed in [f86812d](https://github.com/oxidecomputer/hubris/pull/965/commits/f86812d56e41435275dab8b4cf32982f88d4c35f)